components/tidb_query/src/batch/runner.rs

```diff
@@ -382,29 +395,53 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
                 let mut chunk = Chunk::default();
                 {
                     let data = chunk.mut_rows_data();
-                    data.reserve(
-                        result
-                            .physical_columns
-                            .maximum_encoded_size(&result.logical_rows, &self.output_offsets)?,
-                    );
                     // Although `schema()` can be deeply nested, it is ok since we process data in
                     // batch.
-                    result.physical_columns.encode(
-                        &result.logical_rows,
-                        &self.output_offsets,
-                        self.out_most_executor.schema(),
-                        data,
-                    )?;
+                    match self.encode_type {
+                        EncodeType::TypeDefault => {
+                            data.reserve(result.physical_columns.maximum_encoded_size(
+                                &result.logical_rows,
+                                &self.output_offsets,
+                            )?);
+                            result.physical_columns.encode(
+                                &result.logical_rows,
+                                &self.output_offsets,
+                                self.out_most_executor.schema(),
+                                data,
+                            )?;
+                            chunks.push(chunk);
+                        }
+                        EncodeType::TypeArrow => {
+                            data.reserve(result.physical_columns.maximum_encoded_size_arrow(
+                                &result.logical_rows,
+                                &self.output_offsets,
+                            )?);
+                            result.physical_columns.encode_arrow(
+                                &result.logical_rows,
+                                &self.output_offsets,
+                                self.out_most_executor.schema(),
+                                data,
+                                &self.config.tz,
+                            )?;
+                            batchdata.extend_from_slice(data);
```

Comment:
I remembered that `chunks` lives for a reason that TiDB wants to avoid large bulk of memory. However now for the arrow type all data are put together. I think passing data in the `chunks` field is still needed. For safety considerations, you can add an extra field called `response_format` for TiDB to identify the correct format.